core/jobs/batch_jobs/exp_validation_jobs.py

```diff
@@ -0,0 +1,350 @@
+
+        Returns:
+            bool. Returns whether the exp model is curated or not.
+        """
+        return (model_pair[0] is not None) and (model_pair[1] is not None)
+
+    def get_exploration_from_models(self, model_pair: Tuple[
+        exp_models.ExplorationModel,
+        opportunity_models.ExplorationOpportunitySummaryModel
+    ]) -> exp_domain.Exploration:
+        """Returns the exploration domain object from the curated
+        exploration model.
+
+        Args:
+            model_pair: tuple. The pair of exp and opportunity models.
+
+        Returns:
+            bool. Returns whether the exp model is curated or not.
+        """
+        return exp_fetchers.get_exploration_from_model(model_pair[0])
+
+    def convert_into_model_pair(
+      self, models_dict: ExpAndOpportunityModelsDict
+    ) -> Tuple[
+        Optional[exp_models.ExplorationModel],
+        Optional[opportunity_models.ExplorationOpportunitySummaryModel]
+    ]:
+        """Returns the pair of exp and opportunity models.
+
+        Args:
+            models_dict: dict. The dict of exp and opportunity models list.
+
+        Returns:
+            tuple. The pair of exp and opportunity models.
+        """
+        exp_model = None
+        opportunity_model = None
+        if len(models_dict['exp_models']) == 1:
+            exp_model = models_dict['exp_models'][0]
+        if len(models_dict['opportinity_models']) == 1:
+            opportunity_model = models_dict['opportinity_models'][0]
+        model_pair = (exp_model, opportunity_model)
+        return model_pair
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        """Returns PCollection of details of explorations which are invalid.
+
+        Returns:
+            PCollection[JobRunResult]. A PCollection of details of explorations
+            which are invalid.
+        """
+        all_explorations = (
+            self.pipeline
+            | 'Get all ExplorationModels' >> ndb_io.GetModels(
+                exp_models.ExplorationModel.get_all(include_deleted=False))
+            | 'Create key-value pairs for exp models' >> beam.Map(
+                lambda exp_model: (exp_model.id, exp_model))
+        )
+
+        all_exp_opportunities = (
+            all_explorations
+            | 'Get all ExplorationOpportunitySummaryModels' >>
+                ndb_io.GetModels(
+                    opportunity_models.ExplorationOpportunitySummaryModel
+                        .get_all(include_deleted=False))
+            | 'Create key-value pairs for opportunity models' >> beam.Map(
+                lambda exp_opportunity_model: (
+                    exp_opportunity_model.id, exp_opportunity_model))
+        )
+
+        curated_explorations = (
+            ({
+                'exp_models': all_explorations,
+                'opportinity_models': all_exp_opportunities
+            })
+            | 'Combine the PCollections' >> beam.CoGroupByKey()
```

Comment:
You can also pass tuple to `CoGroupByKey`, that **might** simplify some of the code below.
```suggestion
            (all_explorations, all_exp_opportunities)
            | 'Combine the PCollections' >> beam.CoGroupByKey()
```

Reply:
Done.